cpp/ql/src/semmle/code/cpp/models/implementations/StdPair.qll

```diff
@@ -0,0 +1,61 @@
+/**
+ * Provides models for the C++ `std::pair` class.
+ */
+
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * Additional model for `std::pair` constructors.
+ */
+class StdPairConstructor extends Constructor, TaintFunction {
+  StdPairConstructor() { this.hasQualifiedName("std", "pair", "pair") }
+
+  /**
+   * Gets the index of a parameter to this function that is a reference to
+   * either value type of the pair.
+   */
+  int getAValueTypeParameterIndex() {
+    getParameter(result).getUnspecifiedType().(ReferenceType).getBaseType() =
+      getDeclaringType().getTemplateArgument(_).(Type).getUnspecifiedType() // i.e. the `T1` or `T2` of this `std::pair<T1, T2>`
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // taint flow from second parameter of a value type to the qualifier
+    getAValueTypeParameterIndex() = 1 and
+    input.isParameterDeref(1) and
+    (
+      output.isReturnValue() // TODO: this is only needed for AST data flow, which treats constructors as returning the new object
+      or
+      output.isQualifierObject()
+    )
+  }
+}
+
+/**
+ * An instantiation of `std::make_pair`.
+ */
+class StdMakePair extends TaintFunction {
+  StdMakePair() { this.hasQualifiedName("std", "make_pair") }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    // taint flow from first parameter to the returned object
+    input.isParameterDeref(1) and
```

Comment:
```suggestion
    // taint flow from second parameter to the returned object
    input.isParameterDeref(1) and
```

Reply:
Good spot, I've since removed `StdMakePair` though so this line doesn't exist.